xwiki-platform-core/xwiki-platform-oldcore/src/main/java/org/xwiki/store/hibernate/internal/IDVersionHibernateAdapterFactory.java

```diff
@@ -0,0 +1,168 @@
+/*
+ * See the NOTICE file distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as
+ * published by the Free Software Foundation; either version 2.1 of
+ * the License, or (at your option) any later version.
+ *
+ * This software is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this software; if not, write to the Free
+ * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+ * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+ */
+package org.xwiki.store.hibernate.internal;
+
+import java.sql.DatabaseMetaData;
+import java.sql.SQLException;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.TreeMap;
+
+import javax.inject.Inject;
+import javax.inject.Singleton;
+
+import jakarta.inject.Named;
+
+import org.hibernate.HibernateException;
+import org.hibernate.cfg.Configuration;
+import org.xwiki.component.annotation.Component;
+import org.xwiki.component.descriptor.ComponentDescriptor;
+import org.xwiki.component.manager.ComponentLookupException;
+import org.xwiki.component.manager.ComponentManager;
+import org.xwiki.extension.version.Version;
+import org.xwiki.extension.version.internal.DefaultVersion;
+import org.xwiki.store.hibernate.DatabaseProductNameResolver;
+import org.xwiki.store.hibernate.HibernateAdapter;
+import org.xwiki.store.hibernate.HibernateAdapterFactory;
+
+/**
+ * A helper in charge of finding an adapter based on the database id and version.
+ * <p>
+ * It search for an implementation of {@link HibernateAdapter} with the following role hints:
+ * <ul>
+ * <li>"jdbc scheme/minimum version" (for example the "mysql/8" adapter will be selected for a MySQL 9.0.1 server, if no
+ * adapter is registered for version 9)</li>
+ * <li>"jdbc scheme" (for example the "mysql" adapter will be selected for a MySQL 5.7 server, if no adapter is
+ * registered for version 5.7 or lower)</li>
+ * </ul>
+ * <p>
+ * The version is supposed to mean that the adapter was designed for "this version of greater".
+ * 
+ * @version $Id$
+ */
+@Component
+@Singleton
+@Named("id/version")
+public class IDVersionHibernateAdapterFactory implements HibernateAdapterFactory
+{
+    private static final Version DEFAULT_VERSION = new DefaultVersion("0");
+
+    @Inject
+    private ComponentManager componentManager;
+
+    @Inject
+    private List<DatabaseProductNameResolver> resolvers;
+
+    @Override
+    public Optional<HibernateAdapter> createHibernateAdapter(DatabaseMetaData metaData,
```

Comment:
would have been good to have a unit test for this to cover it

Reply:
It felt better to rely on the integration tests for this, but indeed it's good to test the id/version syntax logic in a unit test even if we have all the cases in the database engines we support.

Reply:
Should have a better coverage now.